File: piwigo/Image/Parameters/Cells/EditImageThumbCollectionViewCell.m
//
//  EditImageThumbCollectionViewCell.m
//  piwigo
//
//  Created by Eddy Lelièvre-Berna on 20/08/2019.
//  Copyright © 2019 Piwigo.org. All rights reserved.
//

#import "CategoriesData.h"
#import "EditImageThumbCollectionViewCell.h"
#import "ImageDetailViewController.h"
#import "ImageService.h"
#import "MBProgressHUD.h"

NSString * const kEditImageThumbCollectionCell_ID = @"EditImageThumbCollectionCell";

@interface EditImageThumbCollectionViewCell() <UITextFieldDelegate>

@property (weak, nonatomic) IBOutlet UIView *imageThumbnailView;
@property (weak, nonatomic) IBOutlet UIImageView *imageThumbnail;
@property (weak, nonatomic) IBOutlet UIView *imageDetails;
@property (weak, nonatomic) IBOutlet UILabel *imageSize;
@property (weak, nonatomic) IBOutlet UILabel *imageFile;
@property (weak, nonatomic) IBOutlet UILabel *imageDate;
@property (weak, nonatomic) IBOutlet UILabel *imageTime;

@property (weak, nonatomic) IBOutlet UIView *editButtonView;
@property (weak, nonatomic) IBOutlet UIButton *editImageButton;

@property (weak, nonatomic) IBOutlet UIView *removeButtonView;
@property (weak, nonatomic) IBOutlet UIButton *removeImageButton;

@property (nonatomic, strong) UIAlertAction *renameFileNameAction;
@property (nonatomic, strong) NSString *oldFileName;

@end

@implementation EditImageThumbCollectionViewCell

-(void)awakeFromNib {
    
    // Initialization code
    [super awakeFromNib];
        
    self.contentView.layer.cornerRadius = 10;
    self.imageThumbnailView.layer.cornerRadius = 14;
    self.imageThumbnail.layer.cornerRadius = 10;
    self.imageDetails.layer.cornerRadius = 10;
    self.editButtonView.layer.cornerRadius = 5;
    self.removeButtonView.layer.cornerRadius = 15;

    self.imageSize.font = [UIFont piwigoFontSmallLight];
    self.imageSize.userInteractionEnabled = NO;

    self.imageDate.font = [UIFont piwigoFontSmallLight];
    self.imageSize.userInteractionEnabled = NO;

    self.imageTime.font = [UIFont piwigoFontSmallLight];
    self.imageTime.userInteractionEnabled = NO;
    
    self.imageFile.font = [UIFont piwigoFontSmallLight];
    self.imageTime.userInteractionEnabled = NO;

    self.editImageButton.tintColor = [UIColor piwigoColorOrange];
    
    // Register palette changes
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(applyColorPalette) name:[PwgNotificationsObjc paletteChanged] object:nil];
}

-(void)applyColorPalette
{
    // Background
    self.imageThumbnailView.backgroundColor = [UIColor piwigoColorBackground];
    self.imageDetails.backgroundColor = [UIColor piwigoColorBackground];
    self.editButtonView.backgroundColor = [UIColor piwigoColorBackground];
    self.removeButtonView.backgroundColor = [UIColor piwigoColorCellBackground];

    // Image size, file name, date and time
    self.imageSize.textColor = [UIColor piwigoColorLeftLabel];
    self.imageFile.textColor = [UIColor piwigoColorLeftLabel];
    self.imageDate.textColor = [UIColor piwigoColorLeftLabel];
    self.imageTime.textColor = [UIColor piwigoColorLeftLabel];
}

-(void)setupWithImage:(PiwigoImageData *)imageData removeOption:(BOOL)hasRemove
{
    // Colors
    [self applyColorPalette];
    
    // Image file name
    self.imageId = imageData.imageId;
    if (imageData.fileName.length > 0) {
        self.imageFile.text = imageData.fileName;
    }
    
    // Show button for renaming file
    [self.editButtonView setHidden:NO];
    
    // Show button for removing image from selection if needed
    if (hasRemove) {
        [self.removeButtonView setHidden:NO];
    } else {
        [self.removeButtonView setHidden:YES];
    }

    // Image from Piwigo server…
    if ((imageData.fullResWidth > 0) && (imageData.fullResHeight > 0)) {
        if (self.bounds.size.width > 299) {     // i.e. larger than iPhone 5 screen width
            self.imageSize.text = [NSString stringWithFormat:@"%ldx%ld pixels, %.2f MB", (long)imageData.fullResWidth, (long)imageData.fullResHeight, (double)imageData.fileSize / 1024.0];
        } else {
            self.imageSize.text = [NSString stringWithFormat:@"%ldx%ld pixels", (long)imageData.fullResWidth, (long)imageData.fullResHeight];
        }
    }

    self.imageDate.text = @"";
    if (imageData.dateCreated != nil) {
        if (self.bounds.size.width > 320) {     // i.e. larger than iPhone 5 screen width
            self.imageDate.text = [NSDateFormatter localizedStringFromDate:imageData.dateCreated dateStyle:NSDateFormatterFullStyle timeStyle:NSDateFormatterNoStyle];
        } else {
            self.imageDate.text = [NSDateFormatter localizedStringFromDate:imageData.dateCreated dateStyle:NSDateFormatterLongStyle timeStyle:NSDateFormatterNoStyle];
        }
        self.imageTime.text = [NSDateFormatter localizedStringFromDate:imageData.dateCreated dateStyle:NSDateFormatterNoStyle timeStyle:NSDateFormatterMediumStyle];
    }

    // Retrieve image thumbnail from Photo Libray
    NSString *thumbnailUrl;
    kPiwigoImageSize albumThumbnailSize = (kPiwigoImageSize)AlbumVars.defaultAlbumThumbnailSize;
    switch (albumThumbnailSize) {
        case kPiwigoImageSizeSquare:
            if (AlbumVars.hasSquareSizeImages) {
                thumbnailUrl = imageData.SquarePath;
            }
            break;
        case kPiwigoImageSizeXXSmall:
            if (AlbumVars.hasXXSmallSizeImages) {
                thumbnailUrl = imageData.XXSmallPath;
            }
            break;
        case kPiwigoImageSizeXSmall:
            if (AlbumVars.hasXSmallSizeImages) {
                thumbnailUrl = imageData.XSmallPath;
            }
            break;
        case kPiwigoImageSizeSmall:
            if (AlbumVars.hasSmallSizeImages) {
                thumbnailUrl = imageData.SmallPath;
            }
            break;
        case kPiwigoImageSizeMedium:
            if (AlbumVars.hasMediumSizeImages) {
                thumbnailUrl = imageData.MediumPath;
            }
            break;
        case kPiwigoImageSizeLarge:
            if (AlbumVars.hasLargeSizeImages) {
                thumbnailUrl = imageData.LargePath;
            }
            break;
        case kPiwigoImageSizeXLarge:
            if (AlbumVars.hasXLargeSizeImages) {
                thumbnailUrl = imageData.XLargePath;
            }
            break;
        case kPiwigoImageSizeXXLarge:
            if (AlbumVars.hasXXLargeSizeImages) {
                thumbnailUrl = imageData.XXLargePath;
            }
            break;

        case kPiwigoImageSizeThumb:
        case kPiwigoImageSizeFullRes:
        default:
            thumbnailUrl = imageData.ThumbPath;
            break;
    }

    if (thumbnailUrl.length <= 0)
    {
        // No known thumbnail URL
        self.imageThumbnail.image = [UIImage imageNamed:@"placeholder"];
        return;
    }
    else
    {
        // Load album thumbnail
        __weak typeof(self) weakSelf = self;
        NSURL *URL = [NSURL URLWithString:thumbnailUrl];
        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:URL];
        [self.imageThumbnail setImageWithURLRequest:request
                                   placeholderImage:[UIImage imageNamed:@"placeholder"]
                                            success:^(NSURLRequest *request, NSHTTPURLResponse *response, UIImage *image) {
                                                weakSelf.imageThumbnail.image = image;
                                            }
                                            failure:^(NSURLRequest *request, NSHTTPURLResponse *response, NSError *error) {
#if defined(DEBUG)
                                                NSLog(@"setupWithImageData — Fail to get thumbnail for image at %@", thumbnailUrl);
#endif
                                            }];
    }
}

-(void)prepareForReuse
{
    [super prepareForReuse];
    
    self.imageFile.text = @"";
    self.imageSize.text = @"";
    self.imageDate.text = @"";
    self.imageTime.text = @"";
}


#pragma mark - Edit Filename

// Propose to edit original filename
-(IBAction)editFileName
{
    // Store old file name
    self.oldFileName = self.imageFile.text;
    
    // Determine the present view controller
    UIViewController *topViewController = [UIApplication sharedApplication].keyWindow.rootViewController;
    while (topViewController.presentedViewController) {
        topViewController = topViewController.presentedViewController;
    }
    
    UIAlertController* alert = [UIAlertController
        alertControllerWithTitle:NSLocalizedString(@"renameImage_title", @"Original File")
                                message:[NSString stringWithFormat:@"%@ \"%@\":", NSLocalizedString(@"renameImage_message", @"Enter a new file name for this image"), self.imageFile.text]
        preferredStyle:UIAlertControllerStyleAlert];
    
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = NSLocalizedString(@"renameImage_title", @"Original File");
        textField.text = self.imageFile.text;
        textField.clearButtonMode = UITextFieldViewModeAlways;
        textField.keyboardType = UIKeyboardTypeDefault;
        textField.keyboardAppearance = AppVars.isDarkPaletteActive ? UIKeyboardAppearanceDark : UIKeyboardAppearanceDefault;
        textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
        textField.autocorrectionType = UITextAutocorrectionTypeYes;
        textField.returnKeyType = UIReturnKeyContinue;
        textField.delegate = self;
    }];

    UIAlertAction* cancelAction = [UIAlertAction
        actionWithTitle:NSLocalizedString(@"alertCancelButton", @"Cancel")
        style:UIAlertActionStyleCancel
        handler:^(UIAlertAction * action) {}];

    self.renameFileNameAction = [UIAlertAction
        actionWithTitle:NSLocalizedString(@"renameCategory_button", @"Rename")
        style:UIAlertActionStyleDefault
        handler:^(UIAlertAction * action) {
            // Rename album if possible
            if(alert.textFields.firstObject.text.length > 0) {
                [self renameImageWithName:alert.textFields.firstObject.text andViewController:topViewController];
            }
        }];
    
    [alert addAction:cancelAction];
    [alert addAction:self.renameFileNameAction];
    alert.view.tintColor = UIColor.piwigoColorOrange;
    if (@available(iOS 13.0, *)) {
        alert.overrideUserInterfaceStyle = AppVars.isDarkPaletteActive ? UIUserInterfaceStyleDark : UIUserInterfaceStyleLight;
    } else {
        // Fallback on earlier versions
    }
    [topViewController presentViewController:alert animated:YES completion:^{
        // Bugfix: iOS9 - Tint not fully Applied without Reapplying
        alert.view.tintColor = UIColor.piwigoColorOrange;
    }];
}

-(void)renameImageWithName:(NSString *)fileName andViewController:(UIViewController *)topViewController
{
    // Display HUD during the update
    [topViewController showPiwigoHUDWithTitle:NSLocalizedString(@"renameImageHUD_label", @"Renaming Original File…") detail:@"" buttonTitle:@"" buttonTarget:nil buttonSelector:nil inMode:MBProgressHUDModeIndeterminate];
    
    // Prepare dictionary of parameters
    NSMutableDictionary *imageInformation = [NSMutableDictionary new];
    [imageInformation setObject:fileName forKey:kPiwigoImagesUploadParamFileName];

    // Rename original filename
    [ImageService setImageFileForImageWithId:self.imageId
                                 withFileName:fileName
                                   onProgress:nil
             OnCompletion:^(NSURLSessionTask *task, NSDictionary *response) {

                if(response != nil)
                    {
                        [topViewController updatePiwigoHUDwithSuccessWithCompletion:^{
                            [topViewController hidePiwigoHUDAfterDelay:kDelayPiwigoHUD completion:^{
                                dispatch_async(dispatch_get_main_queue(), ^{
                                    // Adopt new original filename
                                    self.imageFile.text = fileName;
                                    
                                    // Update parent image view
                                    if ([self.delegate respondsToSelector:@selector(didRenameFileOfImageWithId:andFilename:)])
                                    {
                                        [self.delegate didRenameFileOfImageWithId:self.imageId andFilename:fileName];
                                    }
                                });
                            }];
                        }];
                    }
                    else
                    {
                        [topViewController hidePiwigoHUDWithCompletion:^{
                            [topViewController dismissPiwigoErrorWithTitle:NSLocalizedString(@"renameCategoyError_title", @"Rename Fail") message:NSLocalizedString(@"renameImageError_message", @"Failed to rename your image filename") errorMessage:@"" completion:^{ }];
                        }];
                    }
                } onFailure:^(NSURLSessionTask *task, NSError *error) {
                    [topViewController hidePiwigoHUDWithCompletion:^{
                        [topViewController dismissPiwigoErrorWithTitle:NSLocalizedString(@"renameCategoyError_title", @"Rename Fail") message:NSLocalizedString(@"renameImageError_message", @"Failed to rename your image filename") errorMessage:[error localizedDescription] completion:^{ }];
                    }];
                }];
}


#pragma mark - Remove Image from Selection

-(IBAction)removeImage
{
    // Notify this deselection to parent view
    if ([self.delegate respondsToSelector:@selector(didDeselectImageWithId:)])
    {
        [self.delegate didDeselectImageWithId:self.imageId];
    }
}


#pragma mark - UITextField Delegate Methods

-(BOOL)textFieldShouldBeginEditing:(UITextField *)textField
{
    // Disable Add/Delete Category action
    [self.renameFileNameAction setEnabled:NO];
    return YES;
}

-(BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
    // Enable Rename button if name and extension not empty
    NSString *finalString = [textField.text stringByReplacingCharactersInRange:range withString:string];
    NSString *extension = [finalString pathExtension];
    [self.renameFileNameAction setEnabled:((finalString.length >= 1) && (extension.length >= 3) && ![finalString isEqualToString:self.oldFileName])];
    return YES;
}

-(BOOL)textFieldShouldClear:(UITextField *)textField
{
    // Disable Rename button
    [self.renameFileNameAction setEnabled:NO];
    return YES;
}

-(BOOL)textFieldShouldEndEditing:(UITextField *)textField
{
    return YES;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    return YES;
}

@end

File: piwigo/Image/Parameters/Cells/EditImageThumbCollectionViewCell.swift
//
//  EditImageThumbCollectionViewCell.swift
//  piwigo
//
//  Created by Eddy Lelièvre-Berna on 26/12/2019.
//  Copyright © 2019 Piwigo.org. All rights reserved.
//
//  Converted to Swift 5.3 by Eddy Lelièvre-Berna on 30/08/2021.
//

import Photos
import UIKit
import piwigoKit

@objc protocol EditImageThumbnailDelegate: NSObjectProtocol {
    func didDeselectImage(withID imageID: Int64)
    func didRenameFileOfImage(withId imageID: Int64, andFilename fileName: String)
}

class EditImageThumbCollectionViewCell: UICollectionViewCell
{
    weak var delegate: (any EditImageThumbnailDelegate)?

    @IBOutlet private weak var imageThumbnailView: UIView!
    @IBOutlet private weak var imageThumbnail: UIImageView!
    @IBOutlet private weak var imageDetails: UIView!
    @IBOutlet private weak var imageFile: UILabel!
    @IBOutlet private weak var imageSize: UILabel!
    @IBOutlet private weak var imageFileSize: UILabel!
    @IBOutlet private weak var imageDate: UILabel!
    @IBOutlet private weak var imageTime: UILabel!
    @IBOutlet private weak var editButtonView: UIView!
    @IBOutlet private weak var editImageButton: UIButton!
    @IBOutlet private weak var removeButtonView: UIView!
    @IBOutlet private weak var removeImageButton: UIButton!

    private var imageID = Int64.zero
    private var renameFileNameAction: UIAlertAction?
    private var oldFileName: String?

    override func awakeFromNib() {

        // Initialization code
        super.awakeFromNib()

        contentView.layer.cornerRadius = 10
        imageThumbnailView.layer.cornerRadius = 14
        imageThumbnail.layer.cornerRadius = 10
        imageDetails.layer.cornerRadius = 10
        editButtonView.layer.cornerRadius = 5
        removeButtonView.layer.cornerRadius = 15

        editImageButton.tintColor = PwgColor.orange

        // Register palette changes
        NotificationCenter.default.addObserver(self, selector: #selector(applyColorPalette),
                                               name: Notification.Name.pwgPaletteChanged, object: nil)
    }

    @MainActor
    @objc func applyColorPalette() {
        // Background
        imageThumbnailView.backgroundColor = PwgColor.background
        imageDetails.backgroundColor = PwgColor.background
        editButtonView.backgroundColor = PwgColor.background
        removeButtonView.backgroundColor = PwgColor.cellBackground

        // Image size, file name, date and time
        imageSize.textColor = PwgColor.leftLabel
        imageFile.textColor = PwgColor.leftLabel
        imageFileSize.textColor = PwgColor.leftLabel
        imageDate.textColor = PwgColor.leftLabel
        imageTime.textColor = PwgColor.leftLabel
    }

    func config(withImage imageData: Image?, removeOption hasRemove: Bool) {
        // Colors
        applyColorPalette()
        
        // Check provided image data
        guard let imageData = imageData
        else { return }
        
        // Store image ID
        imageID = imageData.pwgID

        // Image file name
        if imageData.fileName.isEmpty == false {
            imageFile.text = imageData.fileName
            editButtonView.isHidden = false     // Show button for renaming file
        }

        // Show button for removing image from selection if needed
        removeButtonView.isHidden = !hasRemove

        // Image size in pixels, file size
        self.imageSize?.text = imageData.fullRes?.pixels ?? "?x?"
        self.imageFileSize?.text = ByteCountFormatter.string(fromByteCount: imageData.fileSize, countStyle: .file)
        
        // Image date and time
        imageDate.text = ""; imageTime.text = ""
        if imageData.dateCreated != DateUtilities.unknownDateInterval {
            let dateCreated = Date(timeIntervalSinceReferenceDate: imageData.dateCreated)
            let dateFormatter = DateUtilities.dateFormatter
            if bounds.size.width > CGFloat(430) {
                // i.e. larger than iPhone 14 Pro Max screen width
                dateFormatter.dateStyle = .long
                dateFormatter.timeStyle = .none
                imageDate.text = dateFormatter.string(from: dateCreated)
            } else {
                dateFormatter.dateStyle = .short
                dateFormatter.timeStyle = .none
                imageDate.text = dateFormatter.string(from: dateCreated)
            }
            dateFormatter.dateStyle = .none
            dateFormatter.timeStyle = .medium
            imageTime.text = dateFormatter.string(from: dateCreated)
        }

        // Get image from cache or download it
        imageThumbnail.layoutIfNeeded()   // Ensure imageView in its final size
        let scale = max(imageThumbnail.traitCollection.displayScale, 1.0)
        let cellSize = CGSizeMake(imageThumbnail.bounds.size.width * scale, imageThumbnail.bounds.size.height * scale)
        let thumbnailSize = pwgImageSize(rawValue: AlbumVars.shared.defaultAlbumThumbnailSize) ?? .thumb
        PwgSessionDelegate.shared.getImage(withID: imageData.pwgID, ofSize: thumbnailSize, type: .image,
                                           atURL: ImageUtilities.getPiwigoURL(imageData, ofMinSize: thumbnailSize),
                                           fromServer: imageData.server?.uuid) { [weak self] cachedImageURL in
            self?.downsampleImage(atURL: cachedImageURL, to: cellSize)
        } failure: { [weak self] _ in
            self?.setThumbnailWithImage(pwgImageType.image.placeHolder)
        }
    }
    
    private func downsampleImage(atURL fileURL: URL, to cellSize: CGSize) {
        // Downsample image
        let cachedImage = ImageUtilities.downsample(imageAt: fileURL, to: cellSize, for: .image)

        // Set image thumbnail
        setThumbnailWithImage(cachedImage)
    }
    
    private func setThumbnailWithImage(_ image: UIImage) {
        DispatchQueue.main.async {
            self.imageThumbnail.image = image
        }
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()

        imageFile.text = ""
        imageSize.text = ""
        imageDate.text = ""
        imageTime.text = ""
    }

    
    // MARK: - Edit Filename
    // Propose to edit original filename
    @IBAction func editFileName() {
        // Store old file name
        oldFileName = imageFile.text

        // Determine the present view controller
        let topViewController = window?.topMostViewController()

        let alert = UIAlertController(
            title: NSLocalizedString("renameImage_title", comment: "Original File"),
            message: "\(NSLocalizedString("renameImage_message", comment: "Enter a new file name for this image")) \"\(imageFile.text ?? "")\":",
            preferredStyle: .alert)

        alert.addTextField(configurationHandler: { [self] textField in
            textField.placeholder = NSLocalizedString("renameImage_title", comment: "Original File")
            textField.text = imageFile.text
            textField.clearButtonMode = .always
            textField.keyboardType = .default
            textField.keyboardAppearance = AppVars.shared.isDarkPaletteActive ? .dark : .default
            textField.autocapitalizationType = .sentences
            textField.autocorrectionType = .yes
            textField.returnKeyType = .continue
            textField.delegate = self
        })

        let cancelAction = UIAlertAction(
            title: NSLocalizedString("alertCancelButton", comment: "Cancel"),
            style: .cancel,
            handler: { _ in })

        renameFileNameAction = UIAlertAction(
            title: NSLocalizedString("renameCategory_button", comment: "Rename"),
            style: .default,
            handler: { [self] action in
                // Rename album if possible
                if let fileName = alert.textFields?.first?.text, fileName.count > 0 {
                    renameImageFile(withName: fileName,
                                    andViewController: topViewController)
                }
            })

        alert.addAction(cancelAction)
        if let renameFileNameAction = renameFileNameAction {
            alert.addAction(renameFileNameAction)
        }
        alert.view.tintColor = PwgColor.tintColor
        alert.overrideUserInterfaceStyle = AppVars.shared.isDarkPaletteActive ? .dark : .light
        topViewController?.present(alert, animated: true) {
            // Bugfix: iOS9 - Tint not fully Applied without Reapplying
            alert.view.tintColor = PwgColor.tintColor
        }
    }

    @MainActor
    private func renameImageFile(withName fileName: String,
                                 andViewController topViewController: UIViewController?) {
        // Display HUD during the update
        topViewController?.showHUD(withTitle: NSLocalizedString("renameImageHUD_label", comment: "Renaming Original File…"))

        // Prepare parameters for renaming the image/video filename
        let paramsDict: [String : Any] = ["image_id" : imageID,
                                          "file" : fileName,
                                          "single_value_mode" : "replace"]
        // Launch request
        let JSONsession = JSONManager.shared
        JSONsession.postRequest(withMethod: pwgImagesSetInfo, paramDict: paramsDict,
                                jsonObjectClientExpectsToReceive: ImagesSetInfoJSON.self,
                                countOfBytesClientExpectsToReceive: 1000) { result in
            switch result {
            case .success:
                // Filename successfully changed
                DispatchQueue.main.async {
                    topViewController?.updateHUDwithSuccess { [self] in
                        topViewController?.hideHUD(afterDelay: pwgDelayHUD) { [self] in
                            // Adopt new original filename
                            imageFile.text = fileName
                            
                            // Update parent image view
                            delegate?.didRenameFileOfImage(withId: imageID, andFilename: fileName)
                        }
                    }
                }
            
            case .failure(let error):
                /// - Network communication errors
                /// - Returned JSON data is empty
                /// - Cannot decode data returned by Piwigo server
                DispatchQueue.main.async {
                    topViewController?.hideHUD {
                        topViewController?.dismissPiwigoError(
                            withTitle: NSLocalizedString("renameCategoyError_title", comment: "Rename Fail"),
                            message: NSLocalizedString("renameImageError_message", comment: "Failed to rename your image filename"),
                            errorMessage: error.localizedDescription) { }
                    }
                }
            }
        }
    }
    
    
    // MARK: - Remove Image from Selection
    @MainActor
    @IBAction func removeImage() {
        // Notify this deselection to parent view
        delegate?.didDeselectImage(withID: imageID)
    }
}


// MARK: - UITextFieldDelegate Methods
extension EditImageThumbCollectionViewCell: UITextFieldDelegate
{
    func textFieldShouldBeginEditing(_ textField: UITextField) -> Bool {
        // Disable Add/Delete Category action
        renameFileNameAction?.isEnabled = false
        return true
    }

    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        // Enable Rename button if name and extension not empty
        let finalString = (textField.text as NSString?)?.replacingCharacters(in: range, with: string)
        let fileExt = URL(fileURLWithPath: finalString ?? "").pathExtension
        renameFileNameAction?.isEnabled = ((finalString?.count ?? 0) > 0) && ((fileExt.count ) >= 3) && (finalString != oldFileName)
        return true
    }

    func textFieldShouldClear(_ textField: UITextField) -> Bool {
        // Disable Rename button
        renameFileNameAction?.isEnabled = false
        return true
    }

    func textFieldShouldEndEditing(_ textField: UITextField) -> Bool {
        return true
    }

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        return true
    }
}
